server/src/com/thoughtworks/go/config/CachedGoConfig.java

```diff
@@ -18,36 +18,207 @@
+        saveValidConfigToCacheAndNotifyConfigChangeListeners(saveResult.getConfigHolder());
+        return saveResult.getConfigSaveState();
+    }
+
+    public synchronized EntityConfigSaveResult writeEntityWithLock(EntityConfigUpdateCommand updateConfigCommand, Username currentUser) {
+        EntityConfigSaveResult entityConfigSaveResult = dataSource.writeEntityWithLock(updateConfigCommand, this.configHolder, currentUser);
+        saveValidConfigToCacheAndNotifyEntityConfigChangeListeners(entityConfigSaveResult);
+        return entityConfigSaveResult;
+    }
+
+    private <T> void saveValidConfigToCacheAndNotifyEntityConfigChangeListeners(EntityConfigSaveResult<T> saveResult) {
+        saveValidConfigToCache(saveResult.getConfigHolder());
+        LOGGER.info("About to notify {} config listeners", saveResult.getEntityConfig().getClass().getName());
+
+        for (ConfigChangedListener listener : listeners) {
+            if(listener instanceof EntityConfigChangedListener<?> && ((EntityConfigChangedListener) listener).shouldCareAbout(saveResult.getEntityConfig())){
+                try {
+                    long startTime = System.currentTimeMillis();
+                    @SuppressWarnings("unchecked")
+                    EntityConfigChangedListener<T> entityConfigChangedListener = (EntityConfigChangedListener<T>) listener;
+                    entityConfigChangedListener.onEntityConfigChange(saveResult.getEntityConfig());
+                    LOGGER.debug("Notifying {} took (in ms): {}", listener.getClass(), (System.currentTimeMillis() - startTime));
+                } catch (Exception e) {
+                    LOGGER.error("failed to fire config changed event for listener: " + listener, e);
+                }
+            }
+        }
+        LOGGER.info("Finished notifying {} config listeners", saveResult.getEntityConfig().getClass().getName());
+    }
+
+    private synchronized void saveValidConfigToCache(GoConfigHolder configHolder) {
+        if (configHolder != null) {
+            LOGGER.debug("[Config Save] Saving config to the cache");
+            this.lastException = null;
+            this.configHolder = configHolder;
+            this.currentConfig = this.configHolder.config;
+            this.currentConfigForEdit = this.configHolder.configForEdit;
+            serverHealthService.update(ServerHealthState.success(HealthStateType.invalidConfig()));
+        }
+    }
+
+    private synchronized void saveValidConfigToCacheAndNotifyConfigChangeListeners(GoConfigHolder configHolder) {
+        saveValidConfigToCache(configHolder);
+        if (configHolder != null) {
+            notifyListeners(currentConfig);
+        }
+    }
+
+    public String getFileLocation() {
+        return dataSource.getFileLocation();
+    }
+
+    public synchronized void save(String configFileContent, boolean shouldMigrate) throws Exception {
+        GoConfigHolder newConfigHolder = dataSource.write(configFileContent, shouldMigrate);
```

Comment:
Merely moving all methods from `CachedFileGoConfig` to this class "as is". Did not want to change too much as a part of this PR.